components/function-runtimes/nodejs14/kubeless/kubeless_rt/lib/ce.js

```diff
@@ -9,26 +11,30 @@ module.exports = {
 };
 
 function buildEvent(req, res, tracer) {
-    let data = req.body;
-    if (!req.is('multipart/*') && req.body.length > 0) {
-        if (req.is('application/json')) {
-            data = JSON.parse(req.body.toString('utf-8'))
-        } else {
-            data = req.body.toString('utf-8')
-        }
-    }
-
-    return Object.assign( 
-        buildCeHeaders(req), {
+    let event = {
         tracer,
-        data,
         'extensions': { request: req, response: res },
         setResponseHeader: (key, value) => setResponseHeader(res, key, value),
         setResponseContentType: (type) => setResponseContentType(res, type),
         setResponseStatus: (status) => setResponseStatus(res, status),
+        //deprecated
         publishCloudEvent: (data) => publishCloudEvent(data),
-        buildResponseCloudEvent: (id, type, data) => buildResponseCloudEvent(req, id, type, data)
-    });
+        //deprecated
+        buildResponseCloudEvent: (id, type, data) => buildResponseCloudEvent(req, id, type, data),
+        emitCloudEvent: (type, source, data, eventtypeversion) => emitCloudEvent(type, source, data, eventtypeversion),
+    };
+
+    let stringifiedReqestBody = req.body.toString(charset);
+    if (!req.is('multipart/*') && req.body.length > 0) {
+        if(isCloudEvent(req)) {
+            event = Object.assign(event,buildCloudEventAttributes(req, JSON.parse(stringifiedReqestBody)));    
+        } else if (req.is('application/json')) {
```

Comment:
I think that we should wrap the check `req.is('app/json)` in method like the `isCloudEvent`, because we have two different abstraction levels.

Reply:
done